lib/hamt.ml

```diff
@@ -231,36 +230,43 @@ module Make (Config : CONFIG) (Key : Hashtbl.HashedType) :
     | HashCollision (_, _), Leaf (_, _, _) -> combine_tip shift node2 node1
     | _ -> failwith "combine_tip"
 
-  let rec update_list update k = function
-    | [] -> option [] (fun v -> [ (k, v) ]) (update None)
-    | ((kx, vx) as x) :: xs ->
-        if kx = k then option xs (fun v -> (k, v) :: xs) (update (Some vx))
-        else x :: update_list update k xs
+  let update_list =
+    let rec loop update k = function
+      | [] -> ( match update None with None -> [] | Some v -> [ (k, v) ])
+      | ((kx, vx) as x) :: xs ->
+          if kx = k then
+            match update (Some vx) with None -> xs | Some v -> (k, v) :: xs
+          else x :: loop update k xs
+    in
+    fun update k xs -> loop update k xs
```

Comment:
I don't get which closure you're removing here, apart from inling `option`?

Reply:
`(fun v -> (k, v) :: xs)` is the closure need for `option`.

Reply:
But you're right, I don't know why I added this intermediate `loop`. I'll get rid of it.

Reply:
I'll fix in a subsequent PR. If i fix it here, I will create a conflict for myself.